### ISSUE TITLE
MemberNotNullWhen is ignored with static methods

### ISSUE THREAD

**Issue by SpicyCatGames** [dotnet/roslyn#75949]
**Compiler version:** '4.12.0-3.24523.10 (16f58fa3)'. Language version: 13.0.

**Steps to Reproduce:**

```cs
public class Result<T>
{
    [MemberNotNullWhen(true, nameof(Value))]
    public bool IsSuccess { get; }
    public T? Value { get; }

    private Result(T? value, bool isSuccess)
    {
        Value = value;
        IsSuccess = isSuccess;
    }

    public static Result<T> Success(T value) => new(value, true);
    public static Result<T> Failure(string error) => new(default, false);
    [MemberNotNullWhen(true, nameof(Value))]
    public static implicit operator bool(Result<T> result) => result.IsSuccess;
}
```
For the following, `MemberNotNullWhen` seems to be ignored:
```cs
var result = Result<string>.Failure("SomeVal");

if(result.IsSuccess)
    Console.WriteLine(result.Value.ToLower()); // no warning

if (result)
    Console.WriteLine(result.Value.ToLower()); // CS8602 - Dereference of a possibly null reference.
```
**Expected Behavior:**
The compiler doesn't report CS8602 when using result.Value.ToLower()

**Actual Behavior:**
The compiler reports CS8602 when using result.Value.ToLower()

**Comment by RikkiGibson** [comment]
It looks like the goal is to say what when the static method returns true, then a member of parameter `result` is non-null. This is tracked by dotnet/csharplang#4563.

**Comment by jaredpar** [comment]
Closing as external.
